File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.py
def yaksoo(num):
    yak = set()
    
    for i in range(2, int(num**(0.5))+1):
        if num % i == 0:
            yak.add(i)
            yak.add(num // i)
    yak.add(num)
    
    return yak

def cal_max(pos, A, B):
    max_num = 0
    
    for i in pos:
        checkA = True
        checkB = True
        
        for j in range(len(A)):
            if A[j] % i != 0:
                checkA = False
                break
            
            if B[j] % i == 0:
                checkB = False
                break
        
        if not checkA or not checkB:
            continue
            
        max_num = max(max_num, i)
    
    return max_num


def solution(arrayA, arrayB):    
    arrayA.sort()
    arrayB.sort()
    
    minA = arrayA[0]
    minB = arrayB[0]

    maxx = 0
    
    posA = yaksoo(minA)
    posB = yaksoo(minB)
    
    maxx = max(cal_max(posA, arrayA, arrayB), cal_max(posB, arrayB, arrayA))
    
    return maxx

File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.java
import java.util.*;

class Solution {
    public HashSet<Integer> yaksoo(int num) {
        int root = (int) Math.sqrt(num);
        HashSet<Integer> result = new HashSet<>();
        
        for (int i = 2; i < root+1; i++) {
            if (num % i == 0) {
                result.add(i);
                result.add(num/i);
            }
        }
        result.add(num);
        return result;
    }
    
    public int calMax(HashSet<Integer> yaksoo, int[] A, int[] B) {
        int max = 0;
        
        for (int num : yaksoo) {
            Boolean checkA = true;
            Boolean checkB = true;
            
            for (int j : A) {
                if (j % num != 0) {
                    checkA = false;
                    break;
                }
            }
            
            for (int j : B) {
                if (j % num == 0) {
                    checkB = false;
                    break;
                }
            }
            
            if (!checkA || !checkB) {
                continue;
            }
            
            max = Math.max(max, num);
        }
        return max;
    }
    
    public int solution(int[] arrayA, int[] arrayB) {
        Arrays.sort(arrayA);
        Arrays.sort(arrayB);
        
        int minA = arrayA[0];
        int minB = arrayB[0];
        
        HashSet<Integer> posA = yaksoo(minA);
        HashSet<Integer> posB = yaksoo(minB);
        
        return Math.max(calMax(posA, arrayA, arrayB), calMax(posB, arrayB, arrayA));
        
    }
}